### PR TITLE
MINIFICPP-1487 Use a single queue for flow files

### PR DESCRIPTION
**Pull request by fgerlits**
https://issues.apache.org/jira/browse/MINIFICPP-1487

This PR has some overlap with #1025 (the first two commits), so it may be easier to review the commits separately, skipping those.  We may merge either of these PRs, both, or neither.

Instead of a separate `deque` for non-penalized flow files and a `priority_queue` for penalized flow files, this PR uses a single `priority_queue` for both, setting a "due date" to be the insertion time for non-penalized flow files and the penalty expiration time for penalized flow files.

This makes the code in FlowFileQueue simpler, easier to reason about, and the unit tests more robust.  On the other hand, `push` and `pop` are now O(log n) for non-penalized flow files, too, instead of O(1).  I think this is an acceptable cost, as O(log n) is not significantly slower than O(1), but I'm interested in what others think.

---

Thank you for submitting a contribution to Apache NiFi - MiNiFi C++.

In order to streamline the review of the contribution we ask you
to ensure the following steps have been taken:

### For all changes:
- [x] Is there a JIRA ticket associated with this PR? Is it referenced
     in the commit message?

- [x] Does your PR title start with MINIFICPP-XXXX where XXXX is the JIRA number you are trying to resolve? Pay particular attention to the hyphen "-" character.

- [x] Has your PR been rebased against the latest commit within the target branch (typically main)?

- [ ] Is your initial contribution a single, squashed commit?

### For code changes:
- [ ] If adding new dependencies to the code, are these dependencies licensed in a way that is compatible for inclusion under [ASF 2.0](http://www.apache.org/legal/resolved.html#category-a)?
- [ ] If applicable, have you updated the LICENSE file?
- [ ] If applicable, have you updated the NOTICE file?

### For documentation related changes:
- [ ] Have you ensured that format looks appropriate for the output in which it is rendered?

### Note:
Please ensure that once the PR is submitted, you check GitHub Actions CI results for build issues and submit an update to your PR as soon as possible.

### CODE REVIEW

**Comment by adamdebreceni** [comment]
image

!mind the baseline labeled "empty", that is a constant cost of the benchmarking system!

I have measured if there is observable penalty going to O(log n), these tests operate (push/pop) on 10 `std::shared_ptr` instances, popping 10 elements from a 1000000 long priority_queue takes less than a microsecond on my machine, of course compared to the deque it is a slowdown on the order of 10, but based on this, I would not worry about it being a bottleneck of some sort

**Comment by adamdebreceni** [comment]
image


!mind the baseline labeled "empty", that is a constant cost of the benchmarking system!

I have measured if there is observable penalty going to O(log n), these tests operate (push/pop) on 10 `std::shared_ptr<int>` instances (with custom comparator), popping 10 elements from a 1000000 long priority_queue takes less than a `6.5 us` on my machine, of course compared to the deque it is a slowdown by a factor of ~100, but based on this, I would not worry about it being a bottleneck of some sort

**Review by szaszm — APPROVED** [review]

File: libminifi/src/utils/FlowFileQueue.cpp [hunk 1]
```diff
@@ -30,70 +30,41 @@ bool FlowFileQueue::FlowFilePenaltyExpirationComparator::operator()(const value_
...
-  }
-
-  if (!priority_queue_.empty()) {
-    value_type next_flow_file = priority_queue_.top();
-    priority_queue_.pop();
-    return next_flow_file;
-  }
-
-  throw std::logic_error{"forcePop() called on an empty FlowFileQueue"};
+  value_type next_flow_file = queue_.top();
```
Comment (adamdebreceni) [inline libminifi/src/utils/FlowFileQueue.cpp:37]: should we move from `queue_.top()`?

File: extensions/standard-processors/tests/unit/ProcessorTests.cpp [hunk 2]
```diff
@@ -615,65 +615,131 @@ TEST_CASE("TestRPGValid", "[TestRPG6]") {
...
+  incoming_connection_->put(flow_file_2);
+  const auto flow_file_3 = session_->create();
+  incoming_connection_->put(flow_file_3);
+
+  REQUIRE(incoming_connection_->isWorkAvailable());
+  std::set<std::shared_ptr<core::FlowFile>> expired_flow_files;
+  const auto next_flow_file_1 = incoming_connection_->poll(expired_flow_files);
+  REQUIRE(next_flow_file_1 == penalized_flow_file);
+
+  session_->penalize(penalized_flow_file);  // second penalty duration is 200 ms
```
Comment (adamdebreceni) [inline extensions/standard-processors/tests/unit/ProcessorTests.cpp:722]: does this comment apply?

File: libminifi/src/utils/FlowFileQueue.cpp (hunk 1 above)
Comment (fgerlits) [reply libminifi/src/utils/FlowFileQueue.cpp:37]: I don't think we can: `top()` returns a const reference (and is a const member function).

File: extensions/standard-processors/tests/unit/ProcessorTests.cpp (hunk 2 above)
Comment (fgerlits) [reply extensions/standard-processors/tests/unit/ProcessorTests.cpp]: Nope.  Removed in https://github.com/apache/nifi-minifi-cpp/pull/1027/commits/a42c324123487d8a05be45c236ef1b42fd24457e.

File: libminifi/src/utils/FlowFileQueue.cpp (hunk 1 above)
Comment (adamdebreceni) [reply libminifi/src/utils/FlowFileQueue.cpp:37]: that is a bummer, there really is no non-const overload :(

**Review by adamdebreceni — APPROVED** [review]

**Review by lordgamez — APPROVED** [review]

**Pull request closed without merging**